coordinator/src/main/scala/filodb.coordinator/queryplanner/SingleClusterPlanner.scala

```diff
@@ -22,8 +23,68 @@ import filodb.query.LogicalPlan._
 import filodb.query.exec.{LocalPartitionDistConcatExec, _}
 import filodb.query.exec.InternalRangeFunction.Last
 
+// scalastyle:off file.size.limit
+
 object SingleClusterPlanner {
   private val mdNoShardKeyFilterRequests = Kamon.counter("queryengine-metadata-no-shardkey-requests").withoutTags
+
+  // Is TargetSchema changing during query window.
+  private def isTargetSchemaChanging(targetSchemaChanges: Seq[TargetSchemaChange],
+                                     startMs: Long, endMs: Long): Boolean =
+    targetSchemaChanges.nonEmpty && targetSchemaChanges.exists(c => c.time >= startMs && c.time <= endMs)
+
+  // Find the TargetSchema that is applicable i.e effective for the current query window
+  private def findTargetSchema(targetSchemaChanges: Seq[TargetSchemaChange],
+                               startMs: Long, endMs: Long): Option[TargetSchemaChange] = {
+    val tsIndex = targetSchemaChanges.lastIndexWhere(t => t.time <= startMs)
+    if(tsIndex > -1)
+      Some(targetSchemaChanges(tsIndex))
+    else
+      None
+  }
+
+  /**
+   * If TargetSchema exists and all of the target-schema label filters (equals) are provided in the query,
+   * then return true.
+   * @param filters Query Column Filters
+   * @param targetSchema TargetSchema
+   * @return useTargetSchema - use target-schema to calculate query shards
+   */
+  private def useTargetSchemaForShards(filters: Seq[ColumnFilter], targetSchema: Option[TargetSchemaChange]): Boolean =
+    targetSchema match {
+      case Some(ts) if ts.schema.nonEmpty => ts.schema
+        .forall(s => filters.exists(cf => cf.column == s && cf.filter.isInstanceOf[Filter.Equals]))
+      case _ => false
+    }
+
+  /**
+   * Returns an occupied option with target schema labels iff they are defined and identical
+   *   for every leaf LogicalPlan that draws data from a shard (i.e. isn't a scalar).
+   */
+  private def getUniversalTargetSchemaLabels(lp: LogicalPlan,
+                                    tsp: TargetSchemaProvider): Option[Seq[String]] = {
+    lp match {
+      case rs: RawSeries => {
+        val rangeSelectorOpt: Option[(Long, Long)] = rs.rangeSelector match {
+          case IntervalSelector(fromMs, toMs) => Some((fromMs, toMs))
+          case _ => None
+        }
+        val targetSchemaOpt = rangeSelectorOpt.map{ case (fromMs, toMs) =>
+          val tsChanges = tsp.targetSchemaFunc(rs.filters)
+          findTargetSchema(tsChanges, fromMs, toMs).map(_.schema)
+        }
+        if (targetSchemaOpt.isDefined) targetSchemaOpt.get else None
+      }
+      case sc: ScalarPlan => Some(Seq.empty)
+      case nl: NonLeafLogicalPlan => {
+        val tsLabelOpts = nl.children.map(getUniversalTargetSchemaLabels(_, tsp))
+        val allDefinedAndSame = tsLabelOpts.forall(_.isDefined) &&
+          tsLabelOpts.drop(1).forall(_ == tsLabelOpts.head)
```

Comment:
nice trick! some English words to explain to reader what is happening will be useful 🙂

Reply:
Done!